k2/python/tests/expand_ragged_labels_test.py

```diff
@@ -0,0 +1,178 @@
+                        torch.tensor([1, 5, 2, 3, 6, -1],
+                                     dtype=torch.float32,
+                                     device=device)))
+
+                # non-tensor attributes...
+                assert dest.attr1 == src.attr1
+                assert dest.attr2 == src.attr2
+
+                # now for autograd
+                scale = torch.tensor([10, 20, 10, 10, 10, 30], device=device)
+                (dest.float_attr * scale).sum().backward()
+                (dest.scores * scale).sum().backward()
+
+                expected_grad = torch.tensor([10, 20, 30],
+                                             dtype=torch.float32,
+                                             device=device)
+
+                assert torch.all(torch.eq(src.float_attr.grad, expected_grad))
+
+                assert torch.all(torch.eq(src.scores.grad, expected_grad))
+
+    def test_final(self):
+        devices = [torch.device('cpu')]
+        if torch.cuda.is_available():
+            devices.append(torch.device('cuda', 0))
+
+        for device in devices:
+            for need_map in [True, False]:
+                s = '''
+                0 1 2 10
+                0 1 1 20
+                1 2 -1 30
+                2
+                '''
+                src = k2.Fsa.from_str(s).to(device).requires_grad_(True)
+                src.float_attr = torch.tensor([0.1, 0.2, 0.3],
+                                              dtype=torch.float32,
+                                              requires_grad=True,
+                                              device=device)
+                src.int_attr = torch.tensor([1, 2, 3],
+                                            dtype=torch.int32,
+                                            device=device)
+                src.ragged_attr = k2.RaggedInt('[[1 2 3] [5 6] [1]]').to(device)
+
+
+                src.attr1 = 'src'
+                src.attr2 = 'fsa'
+
+                if need_map:
+                    dest, arc_map = k2.expand_ragged_labels(src, ret_arc_map = True)
+                else:
+                    dest = k2.expand_ragged_labels(src)
+
+                assert torch.allclose(
+                    dest.float_attr,
+                    torch.tensor([0.1, 0.2, 0.0, 0.0, 0.0, 0.3, 0.0],
+                                 dtype=torch.float32,
+                                 device=device))
+
+
+                assert torch.all(
+                    torch.eq(
+                        dest.scores,
+                        torch.tensor([10, 20, 0, 0, 0, 30, 0],
+                                     dtype=torch.float32,
+                                     device=device)))
+
+                assert torch.all(
+                    torch.eq(
+                        dest.int_attr,
+                        torch.tensor([1, 2, 0, 0, 0, 3, 0], dtype=torch.int32, device=device)))
+                _k2.fix_final_labels(dest.arcs, dest.int_attr)
```

Comment:
Do we need to add a note in the function `k2.expand_ragged_labels`
saying that the user may need to call `_k2.fix_final_labels`?

BTW, `_k2.fix_final_labels` is not exported to the `k2` module in Python.

Reply:
Or as another alternative, can we add an extra argument to `_k2.expand_ragged_labels`
to accept also a list of normal tensors. We can change them in a single kernel, which is more efficient.
And it is less error-prone as there is no need for the user to call `_k2.fix_final_labels`.

Reply:
Regarding needing to call _k2.fix_final_labels:
  The reason I'm not mentioning this is that we don't have a general requirement for attributes, that for final arcs they must have the value -1.  I am enforcing this in this case because it seems likely to be what would be wanted, but I may change my mind about this in future and let them be zero (as I think I envisaged at one point early on).
Here, I just wanted to test the function; I don't have an immediate use for it.